Authentication using HTTP Client Certificates

sklirg opened the pull request:
Fixes #40, fixes #41

sklirg committed:
fix: Correct log level and add errors to some log lines

sklirg committed:
chore: Add CA and Client Cert generation

sklirg committed:
feat: Implement support for Client Certificates in HTTP sender and receiver.

Fixes #41

sklirg commented on receiver/http.go:
```diff
@@ -148,6 +168,19 @@ func (htt *HTTP) Start() error {
+		}
+		server.TLSConfig = &tls.Config{
+			ClientCAs:  pool,
+			ClientAuth: tls.RequireAndVerifyClientCert,
```
Should the policy of authentication be configurable? If so, which auth types should we support?

https://golang.org/pkg/crypto/tls/#ClientAuthType

KristianLyng replied:
See previous comment, but I think "nyeso".

It makes very little sense to accept and/or require Client certificates without verifying them, so the only valid setting I can imagine is disabling it completely (the default) or RequireAndVerifyClientCert.

sklirg committed:
feat: Verify Client Certificate based on the presence of an expected SAN DNS Name

KristianLyng commented on receiver/http.go:
```diff
@@ -133,6 +137,22 @@ func (rcvr receiver) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+// loadClientCertificateCAs loads a given list of strings as paths of
+// acceptable CAs for use in Client Certificate authentication.
+func loadClientCertificateCAs(paths []string) (*x509.CertPool, error) {
+	httpLog.Debugf("Loading %d Client Certificate(s) from file(s)", len(paths))
```
This is probably incorrect, isn't it? It's common to store multiple certificates in a single file (my /etc/ssl/certs/ca-certificates.crt has 127 ca certificates). So the message should probably read "Loading %d certificate files" or something like that. Perhaps somewhat nitpicky, but it can certainly create confusion during troubleshooting.

KristianLyng commented on receiver/http.go:
```diff
@@ -47,11 +50,12 @@ least one handler to be set up, using Handle. This is done implicitly
+	Auth          map[string]*HTTPAuth          `doc:"A map corresponding to Handlers; specifying authentication for the given path, if required."`
+	Certfile      string                        `doc:"Path to certificate file for TLS. If left blank, un-encrypted HTTP is used."`
+	Keyfile       string                        `doc:"Path to key file for TLS."`
+	AcceptableCAs []string                      `doc:"Paths to files containing CAs which are accepted for Client Certificate authentication."`
```
I'm not sure I like this naming. And I'm curious if we can also use the default system-wide CA? In that case, we could/would need two options. One to enable client certificate validation, the other to list custom CAs... That might be overkill though. Thoughts?

As for the name, regardless of if it's one or two: What about ClientCertificateCAs ? The reason I'm skeptical to AcceptableCAs is that client certificates is somewhat obscure to most people and as such just saying "AcceptableCAs" could mislead people to think it's about server-side certificates.

KristianLyng commented on sender/http.go:
```diff
@@ -95,13 +97,22 @@ func getCertPool(path string) (*x509.CertPool, error) {
 	}
 	if ht.Insecure {
-		httpLog.Print("Warning: Disabeling certificate validation for HTTP sender - vulnerable to man-in-the-middle")
+		httpLog.Warning("Disabling certificate validation for HTTP sender - vulnerable to man-in-the-middle")
```
Nice catch :D

KristianLyng commented on sender/http.go:
```diff
@@ -111,15 +122,29 @@ func (ht *HTTP) init() {
+		RootCAs:            cp,
+	}
+
+	if ht.Certfile != "" && ht.Keyfile != "" {
```
Since both are required if either is specified, we should probably check for this in Verify() to make sure.

KristianLyng reviewed (changes requested):
I am fairly happy with this, but I haven't actually tested it. It seems to do what it needs to do.

I am honestly not sure if having two options in the receiver instead of one (see discussion) is better, so I'll leave that decision to you. It can always be introduced later on. Come to think of it, since we're unsure, I think just having the simplest implementation (the one you suggest) is the best way to move forward.

Other than that, I would prefer a name change for AcceptableCAs, but it's not a requirement.

However, for the sender, we should Verify() that either both key/cert is provided, or neither, to avoid user-confusion. Since that's trivial, that's the only thing I'm really making a requirement from my thumbs up.

sklirg committed:
chore: Use better name for CAs for client certificates in config

sklirg committed:
chore: Explain that we're searching \%d files, not \%d certs

sklirg committed:
feat: Verify that we have both certfile and keyfile if either is specified

sklirg committed:
feat: Engage VerifyPeerCertificate only on configured paths

Also, if a Client Certificate is presented upon connecting, we verify
the certificate to be valid for one of the paths published.
